### ISSUE TITLE
Spacing in closures

### ISSUE THREAD

**Issue by mboynes** [WordPress-Coding-Standards/WordPress-Coding-Standards#439]
Many of my Travis-tested repos just started failing due to the merge of #417. I see that WPCS now wants:

```php
add_filter( 'excerpt_length', function() {
	return 40;
} );
```

to be:

```php
add_filter( 'excerpt_length', function () {
	return 40;
} );
```

I see no mention of this in the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/). Is there a secondary guide that WPCS is using to make decisions like this? If it's a matter of opinion, I'd love to see the discussion so I can better understand why this format was chosen. I personally think `function() {` makes more, as it follows the non-closure function standard more closely, where the parens are adjacent to the function name like `function foo() {`. If it were just a matter of opinion, I would hide it in my ruleset, though it doesn't look like this can be ignored without also affecting other control structures (please correct me if I'm wrong) covered by `WordPress.WhiteSpace.ControlStructureSpacing.NoSpaceAfterStructureOpen` and `WordPress.WhiteSpace.ControlStructureSpacing.NoSpaceBeforeOpenParenthesis`.

Thanks!

**Comment by lkraav** [comment]
Yes, this is ouch. I was just about to file the same issue.

**Comment by GaryJones** [comment]
I raised this at https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/pull/417#discussion_r35848698 . In this case, I can only assume that @JDGrimes defaulted (or let the Squiz sniff default) to PSR-2, which does have spaces: http://www.php-fig.org/psr/psr-2/#6-closures

Anonymous functions / closures tend not to be found in Core, so things like this, namespaces and other non-core goodies are something that I've raised with @DrewAPicture so that we can get it clarified in the handbooks regardless of whether WP core is using it.

Drew, can this particular point be raised sooner with the Core team, so we can amend as necessary? FWIW, no-space (like WP CS JS uses) is my slight preference. The only occurrence I could see in core was in tests/phpunit/actions/closures where no-space is used twice. If this is sufficiently indicative, then a change will be needed in WPCS.

**Comment by JDGrimes** [comment]
I don't have a strong preference either way, but I tend to use the space. ("When in doubt, space it out!" :smiling_imp:) But we can make `0` the default, or even have it default to no preference until core makes a decision.

**Comment by GaryJones** [comment]
The fact that 100% of instances have no-space in core (OK, so 2 out of 2, in tests (!)), named function declarations don't have a space, and anonymous functions in JS don't (and there being some intent to have consistency between JS and PHP), I think I'd prefer no-space to be expected for `function()`, until a decision is made otherwise.

**Comment by JDGrimes** [comment]
OK, should we release a 0.7.1 with that change?

**Comment by westonruter** [comment]
@JDGrimes great idea :+1:

**Comment by GaryJones** [comment]
+1 for 0.7.1. I meant to suggest that in my previous comment.

**Comment by mboynes** [comment]
Thanks for the quick turnaround!

**Comment by schlessera** [comment]
I'm not happy with the reasoning behind this decision, as I would argue that it does not make sense at all.

A named function declaration would be `function whatever( $item ) { [...] }`.

In a closure, what gets omitted is the name of the function, so you'd remove the `whatever` above. This leaves you with `function ( $item ) { [...] }`.

You're NOT passing arguments to `function`, you're declaring an unnamed entity of type `function` with so and so arguments, just like you would have `int 5` instead of `int5`...

**Comment by GaryJones** [comment]
@schlessera also pointed out that PHPStorm doesn't even have the Coding Style option (yet?) to remove that space.

**Comment by schlessera** [comment]
Haha, apparently they consider this to be a bug in PHPStorm...

Oh well, nevertheless, I maintain my arguments, it is not a call to `function` with arguments, it is `function <no-name-here>( $item ) {}`...

**Comment by JDGrimes** [comment]
@schlessera I agree, which is why I use the space. But this project doesn't control the standards that WordPress uses, we just implement sniffs for those standards. We need to be compatible with WordPress's declared standards by default, which is why this change was made. Those of us that disagree with [WordPress's standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/) can try to get them changed. Until that happens, you can override the default in your phpcs.xml config:

```xml
<rule ref="WordPress.WhiteSpace.ControlStructureSpacing">
    <properties>
        ...
        <property name="spaces_before_closure_open_paren" value="1" />
    </properties>
</rule>
```

**Comment by schlessera** [comment]
Yes, I agree with that, but in this case, WordPress does not have any standard to speak of. The two instances that @GaryJones managed to find can't possibly be considered conclusive.

So, why not just ignore this until Core has made a decision?

Yes, I'll consider overriding it in my projects. Thanks!

**Comment by JDGrimes** [comment]
> Yes, I agree with that, but in this case, WordPress does not have any standard to speak of. The two instances that @GaryJones managed to find can't possibly be considered conclusive.

No, those instances aren't conclusive. I would have favored turning off this check entirely by default, except that it is a part of the [JavaScript standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/), and a strong emphasis has been placed on parity between the PHP and JS standards in regards to whitespace.

I'm still not against ignoring this check by default until it is clarified in the PHP standards. However, that will mean that _everyone_ will need to implement a custom configuration if they want to sniff for that. @GaryJones @westonruter any thoughts?

**Comment by westonruter** [comment]
I think it should be ignored by default. It can be an opt-in for now, and once Core has a standard for closure formatting, we can enable the check.

**Comment by GaryJones** [comment]
As per 24 days ago:

> I think I'd prefer no preference to be expected for function(), until a decision is made otherwise.
----
> Those of us that disagree with WordPress's standards can try to get them changed
Unfortunately, there isn't a Slack channel or other method that can deal with this, without conversations getting lost in core chat.

@DrewAPicture has taken my suggestions (raised in Docs chat) about the PHP CS listed in the Handbooks to core devs, but really, the choosing of standards for use in the community should be a separate group. They'd (core and non-core devs) decide on the standards (with strong influence from historical use in core, obviously), the standards are written up via the Docs team, given to WPCS to be programatically testable, and then core, plugin and theme authors could use them. This lack of the initial CS group is just frustrating.

**Comment by JDGrimes** [comment]
> As per 24 days ago:
> 
>     I think I'd prefer no preference to be expected for function(), until a decision is made otherwise.

Ah, I must have missed that somehow. I don't know why I defaulted it to `0`. (But don't blame me, _you're_ the one who merged it :smile:).

**Comment by GaryJones** [comment]
> But don't blame me, you're the one who merged it

I deferred to your expert decision :-)
